File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    genre_songs = {} 
    genre_plays = {}  
    
    for i, (genre, play) in enumerate(zip(genres, plays)):
        if genre not in genre_songs:
            genre_songs[genre] = [[i, play]]
            genre_plays[genre] = play
        else:
            genre_songs[genre].append([i, play])
            genre_plays[genre] += play
    
    sorted_genres = sorted(genre_plays.keys(), key=lambda x: -genre_plays[x])
    
    for genre in sorted_genres: 
        sorted_songs = sorted(genre_songs[genre], key=lambda x: -x[1])[:2]
        for song in sorted_songs:
            answer.append(song[0])
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

public class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, List<int[]>> genreSongs = new HashMap<>();
        Map<String, Integer> genrePlays = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {
            String genre = genres[i];
            int play = plays[i];

            genreSongs.putIfAbsent(genre, new ArrayList<>());
            genreSongs.get(genre).add(new int[]{i, play});

            genrePlays.put(genre, genrePlays.getOrDefault(genre, 0) + play);
        }

        List<String> sortedGenres = new ArrayList<>(genrePlays.keySet());
        sortedGenres.sort((a, b) -> genrePlays.get(b) - genrePlays.get(a));

        List<Integer> answer = new ArrayList<>();

        for (String genre : sortedGenres) {
            List<int[]> songs = genreSongs.get(genre);

            songs.sort((a, b) -> {
                if (b[1] == a[1]) return a[0] - b[0];
                return b[1] - a[1];
            });

            for (int i = 0; i < Math.min(2, songs.size()); i++) {
                answer.add(songs.get(i)[0]);
            }
        }

        return answer.stream().mapToInt(i -> i).toArray();
    }
}
